tests/unit/test_charm.py

```diff
@@ -3,18 +3,55 @@
 #
 # Learn more about testing at: https://juju.is/docs/sdk/testing
 
-import unittest
-
 import ops
 import ops.testing
-from charm import OpenstackExporterOperatorCharm
+import pytest
+from charm import CLOUD_NAME, OS_CLIENT_CONFIG, SNAP_NAME, OpenstackExporterOperatorCharm
 
 
-class TestCharm(unittest.TestCase):
-    def setUp(self):
+class TestCharm:
+    def setup_method(self, _):
         self.harness = ops.testing.Harness(OpenstackExporterOperatorCharm)
-        self.addCleanup(self.harness.cleanup)
+
+    def teardown_method(self, _):
+        self.harness.cleanup()
+
+    @pytest.mark.parametrize(
+        "config",
+        [
+            ({"cache": False, "cache_ttl": "100s"}),
+            ({"cache": True, "cache_ttl": "200s"}),
+        ],
+    )
+    def test_config_changed(self, config, mocker):
+        mock_get_installed_snap_service = mocker.patch("charm.get_installed_snap_service")
+        mock_snap_service = mocker.Mock(spec_set=["configure", "restart_and_enable", "stop"])
+        mock_get_installed_snap_service.return_value = mock_snap_service
+
         self.harness.begin()
 
-    def test_null(self):
-        """Remove me."""
+        mock_get_keystone_data = mocker.Mock()
+        mock_keystone_data = mocker.Mock()
+        mock_get_keystone_data.return_value = mock_keystone_data
+        self.harness.charm._get_keystone_data = mock_get_keystone_data
+
+        mock_write_cloud_config = mocker.Mock()
+        self.harness.charm._write_cloud_config = mock_write_cloud_config
```

Comment:
Why not using `self.harness.charm._write_cloud_config.assert_called_with(...)`, why we need to create mock object and call `self.harness.charm._write_cloud_config = ...`? Should we rather properly mock this object and not change it behavior with assigning something else, using `obj = ...` instead of `path.object(self.harness.charm, "_write_cloud_config")`?

Reply:
I think we don't care the details of `_write_cloud_config` in this unit test case. We only make sure it's been called with right arguments.